cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,232 @@
+    "cluster-autoscaler.kubernetes.io/consume-provisioning-request": "provreq-name"
+```
+
+Note: CA will match all pods with this annotation to a corresponding ProvReq and
+ignore them when executing a scale-up loop (so that is up to users to make sure
+that the ProvReq count is matching the number of created pods).
+
+### CRD lifecycle
+
+1.  A ProvReq will be created either by the end user or by a framework.
+2.  CA will pick it up, choose a nodepool (or create a new one if NAP is
+    enabled), and try to create nodes.
+3.  If CA successfully creates capacity, ProvReq will receive information about
+    this fact in `Conditions` field.
+4.  At this moment, users can create pods in that will consume the ProvReq (in
+    the same namespace), those will be scheduled on the capacity that was
+    created by the CA.
+5.  Once all of the pods are scheduled users can delete the ProvReq object,
+    otherwise it will be garbage collected after some time.
+6.  When pods finish the work and nodes become unused the CA will scale them
+    down.
+
+Note: Users can create a ProvReq and pods consuming them at the same time (in a
+"fire and forget" manner), but this may result in the pods being unschedulable
+and triggering user configured alerts.
+
+### Conditions
+
+The following Condition states should encode the states of the ProvReq:
+
+  - Provisioned - VMs were created successfully (Atomic class)
+  - CapacityAvailable - cluster contains enough capacity to schedule pods (Check
+    class)
+  - NotEnoughCapacity - not enough capacity in cluster (Check class)
+  - Failed - failed to create or check capacity (both classes)
+
+The Reasons and Messages will contain more details about why the specific
+condition was triggered.
+
+### CA implementation details
+
+The proposed implementation is to handle each ProvReq in a separate scale-up
+loop. This will require changes in multiple parts of CA:
+
+1.  Listing unschedulable pods where:
+      - pods that consume ProvReq need to filtered-out
+      - pods that are represented by the ProvReq need to be injected (we need to
+        ensure those are threated as one group by the sharding logic)
+2.  Scale-up logic, which as of now has no notion atomicity and grouping of
+    pods. This is simplified as the ScaleUp logic was recently put [behind an
+    interface](https://github.com/kubernetes/autoscaler/pull/5597).
+      - This is a place where the biggest part of the change will be made. Here
+        many parts of the logic are assuming best-effort semantics and the scale
+        up size is lowered in many situations:
+          - Estimation logic, which stops after some time-out or number of
+            pods/nodes.
+          - Size limiting, which caps the scale-up to match the size
+            restrictions (on node group or cluster level).
+3.  Node creation, which needs to support atomic resize. Either via native cloud
+    provider APIs or best effort with node removal if CA is unable to fulfill
```

Comment:
Best effort atomic could be:
1. single attempt
2. retries -- scaling down to zero and retrying later
3. staircase -- holding what we got and trying further

Which one are we planning?

Reply:
2. Retries

I've updated the doc to describe this in more details.

Reply:
Would you consider a staircase implementation for day 2? This would be beneficial for gang-scheduling use cases that require difficult to acquire machine types (e.g. GPUs).

Reply:
The classes listed here are examples of what can be supported in CA, a class where we provide VMs in a staircase is definitely possible.
Though, I think the current CA behaviour would a very similar to it, i.e. if you have unschedulable pods CA will try to create a nodes in a back-off mode, yielding a staircase like behaviour.